packages/react-aria-components/src/Table.tsx

```diff
@@ -466,27 +476,69 @@ export interface TableHeaderProps<T> extends StyleProps {
   dependencies?: any[]
 }
 
-function TableHeader<T extends object>(props: TableHeaderProps<T>, ref: ForwardedRef<HTMLTableSectionElement>) {
-  let children = useCollectionChildren({
+/**
+ * A header within a `<Table>`, containing the table columns.
+ */
+export const TableHeader =  /*#__PURE__*/ createBranchComponent(
+  'tableheader',
+  <T extends object>(props: TableHeaderProps<T>, ref: ForwardedRef<HTMLTableSectionElement>) => {
+    let collection = useContext(TableStateContext)!.collection as TableCollection<unknown>;
+    let headerRows = useCachedChildren({
+      items: collection.headerRows,
+      children: useCallback((item: Node<unknown>) => {
+        switch (item.type) {
+          case 'headerrow':
+            return <TableHeaderRow item={item} />;
+          default:
+            throw new Error('Unsupported node type in TableHeader: ' + item.type);
+        }
+      }, [])
+    });
+  
+    let {rowGroupProps} = useTableRowGroup();
+    return (
+      <thead
+        {...filterDOMProps(props as any)}
+        {...rowGroupProps}
+        ref={ref}
+        className={props.className ?? 'react-aria-TableHeader'}
+        style={props.style}>
+        {headerRows}
+      </thead>
+    );
+  },
+  props => useCollectionChildren({
```

Comment:
What kinds of use case requests were there for custom collection components? We'll need to expose `useCollectionChildren` and `CollectionContext` for cases similar to TableHeader and TableRow which render other collection components. Feels like a pretty advanced use case but not sure if there is a good way to simplify it

Reply:
yeah we'd expose those along with the `createLeafComponent` API etc.